src/test/java/io/cdap/plugin/servicenow/source/ServiceNowSourceConfigTest.java

```diff
@@ -0,0 +1,426 @@
+import static io.cdap.plugin.servicenow.source.ServiceNowSourceConfigHelper.TEST_CLIENT_ID;
+import static io.cdap.plugin.servicenow.source.ServiceNowSourceConfigHelper.TEST_CLIENT_SECRET;
+import static io.cdap.plugin.servicenow.source.ServiceNowSourceConfigHelper.TEST_PASSWORD;
+import static io.cdap.plugin.servicenow.source.ServiceNowSourceConfigHelper.TEST_USER;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_API_ENDPOINT;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_APPLICATION_NAME;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_CLIENT_ID;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_CLIENT_SECRET;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_END_DATE;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_PASSWORD;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_QUERY_MODE;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_START_DATE;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_TABLE_NAME;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_TABLE_NAME_FIELD;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_USER;
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PROPERTY_VALUE_TYPE;
+
+/**
+ * Tests for {@link ServiceNowSourceConfig}.
+ */
+public class ServiceNowSourceConfigTest {
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void testQueryModeTable() {
+    SourceQueryMode queryMode = SourceQueryMode.TABLE;
+    ServiceNowSourceConfig config = ServiceNowSourceConfigHelper.newConfigBuilder()
+      .setQueryMode("Table")
+      .build();
+
+    MockFailureCollector collector = new MockFailureCollector();
+    Assert.assertEquals(queryMode, config.getQueryMode(collector));
+    Assert.assertEquals(0, collector.getValidationFailures().size());
+  }
+
+  @Test
+  public void testQueryModeReporting() {
+    SourceQueryMode queryMode = SourceQueryMode.REPORTING;
+    ServiceNowSourceConfig config = ServiceNowSourceConfigHelper.newConfigBuilder()
+      .setQueryMode("Reporting")
+      .build();
+
+    MockFailureCollector collector = new MockFailureCollector();
+    Assert.assertEquals(queryMode, config.getQueryMode(collector));
+    Assert.assertEquals(0, collector.getValidationFailures().size());
+  }
+
+  @Test
+  public void testQueryModeInvalid() {
+    ServiceNowSourceConfig config = ServiceNowSourceConfigHelper.newConfigBuilder()
+      .setQueryMode(null)
+      .build();
+
+    try {
+      MockFailureCollector collector = new MockFailureCollector();
+      config.getQueryMode(collector);
+    } catch (ValidationException e) {
+      Assert.assertEquals(1, e.getFailures().size());
```

Comment:
This test case will pass (incorrectly) if config.getQueryMode() does not throw ValidationException when query mode is invalid. Similar comment for other test cases that expect ValidationException to be thrown.

Reply:
Done

Reply:
How was this addressed?

Reply:
We are asserting size of validation failures in collector outside the catch block